lxd/instance/drivers/docs/vm-live-migration-developer-documentation.md

```diff
@@ -0,0 +1,186 @@
+# LXD VM Live Migration
+
+LXD supports [live migration](https://documentation.ubuntu.com/lxd/en/latest/howto/move_instances/#live-migration) of virtual machines by streaming instance state from a source QEMU to a target QEMU. VM live migration is supported for all storage pool types.
+
+Please see the [specification](https://discuss.linuxcontainers.org/t/lxd-online-vm-live-migration-qemu-to-qemu/16635) for further implementation details, design choices, and relevant pull requests.
+
+API Extension: `migration_vm_live`
+
+## Conceptual Process
+
+The live migration workflow varies depending on the type of storage pool used. The two key scenarios are non-shared storage and shared storage within a cluster (e.g., Ceph). If live state transfer is not supported by a target, a stateful stop is performed prior to migration.
+
+### Non-Shared Storage Live Migration
+
+This process leverages QEMU’s built-in Network Block Device (NBD) client-server mechanism to transfer the VM’s disk and memory state. Below is an overview of the steps:
+
+1. Set QMP Capabilities.
+2. Temporary Write Buffer: Create a temporary QCOW2 file to store writes that occur in the guest during the initial transfer of the root disk.
+3. Snapshot Root Disk.
+4. DB Record Creation: Create the target instance and volume DB records.
+5. Storage Driver Transfer.
+6. Start Target QEMU: Launch the QEMU process on the target.
+7. Add NBD Node: Attach the NBD target disk as a block node on the source. 
+8. Block Device Mirror: Perform a block device mirror of the snapshot to the target NBD server.
+9. Pause Source VM: Pause the source VM once the block device mirror enters `ready` state.
+10. State Transfer: Transfer the memory state over websocket.
+11. Enter `pre-switchover` State.
+12. Complete Migration: Continue state migration until `completed` state reached.
+
+```mermaid
+---
+title: LXD VM Live Migration - Non-Shared Storage Process
+---
+flowchart TD
+    Start[Start Migration] --> ValidateStorage[Validate Non-Shared Storage]
+    ValidateStorage -->|Storage Valid| PrepareSnapshot[Create Snapshot of Root Disk]
+    ValidateStorage -->|Storage Invalid| Error[Abort Migration]
+    PrepareSnapshot --> ConfigureWrites[Redirect Disk Writes to Snapshot]
+    ConfigureWrites --> SetupTarget[Set Up Target Instance]
+    SetupTarget --> TransferStorage[Transfer Root Disk Using NBD]
+    TransferStorage --> StartTarget[Start Target QEMU Process]
+    StartTarget --> SyncMemory[Transfer and Sync Memory State]
+    SyncMemory --> PauseVM[Pause Source VM Guest]
+    PauseVM --> PerformFinalSync[Perform Final Memory Sync]
+    PerformFinalSync --> SwitchTarget[Switch VM to Target Host]
+    SwitchTarget --> Cleanup[Clean Up Resources]
+    Cleanup --> Complete[Migration Complete]
+```
+
+The state transitions during the process are shown below:
+
+```mermaid
```

Comment:
We are putting @markylaing OVN docs into the LXD Internals section of our docs, so we should use SVG diagrams I understand. See https://github.com/canonical/lxd/pull/14556